crates/transaction-pool/src/maintain.rs

```diff
@@ -0,0 +1,268 @@
+                                new_tip.hash
+                            );
+                            dirty.extend(addresses);
+                            vec![]
+                        }
+                    };
+
+                // also include all accounts from new chain
+                // we can use extend here because they are unique
+                changed_accounts.extend(new_changed_accounts.into_iter().map(|entry| entry.0));
+
+                // all transactions mined in the new chain
+                let new_mined_transactions: HashSet<_> =
+                    new_blocks.transactions().map(|tx| tx.hash).collect();
+
+                // update the pool then reinject the pruned transactions
+                // find all transactions that were mined in the old chain but not in the new chain
+                let pruned_old_transactions = old_blocks
+                    .transactions()
+                    .filter(|tx| !new_mined_transactions.contains(&tx.hash))
+                    .filter_map(|tx| tx.clone().into_ecrecovered())
+                    .map(|tx| {
+                        <V as TransactionValidator>::Transaction::from_recovered_transaction(tx)
+                    })
+                    .collect();
+
+                // update the pool first
+                let update = CanonicalStateUpdate {
+                    hash: new_tip.hash,
+                    number: new_tip.number,
+                    pending_block_base_fee,
+                    changed_accounts,
+                    // all transactions mined in the new chain need to be removed from the pool
+                    mined_transactions: new_mined_transactions.into_iter().collect(),
+                };
+                pool.on_canonical_state_change(update);
+
+                // all transactions that were mined in the old chain but not in the new chain need
+                // to be re-injected
+                //
+                // Note: we no longer know if the tx was local or external
+                let _ = pool.add_external_transactions(pruned_old_transactions).await;
+                // TODO: metrics
+            }
+            CanonStateNotification::Revert { old } => {
+                // this similar to the inverse of a commit where we need to insert the transactions
+                // back into the pool and update the pool's state accordingly
+
+                let (blocks, state) = old.inner();
+                let first_block = blocks.first();
+                if first_block.hash == pool_info.last_seen_block_hash {
+                    // nothing to update
+                    continue
+                }
+
+                // base fee for the next block: `first_block+1`
+                let pending_block_base_fee = calculate_next_block_base_fee(
+                    first_block.gas_used,
+                    first_block.gas_limit,
+                    first_block.base_fee_per_gas.unwrap_or_default(),
+                ) as u128;
```

Comment:
irrelevant to this PR: would be nice to declare this fn on `Header` type.
```rust
let next_base_fee = block.next_base_fee();
```

Reply:
yeh I'd like that as well, having this in the consensus common crate is not really ideal, will track in separate issue